jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdCredentials.java

```diff
@@ -224,40 +223,23 @@ and the Token Endpoint (which it is in this flow), the TLS server validation
         return paddedEncodedJwtSection;
     }
 
-    private Map<String, Object> claimAuthCode(String authCode) throws IOException
+    private Map<String, Object> claimAuthCode(HttpClient httpClient, String authCode) throws Throwable
     {
+        Fields fields = new Fields();
+        fields.add("code", authCode);
+        fields.add("client_id", configuration.getClientId());
+        fields.add("client_secret", configuration.getClientSecret());
+        fields.add("redirect_uri", redirectUri);
+        fields.add("grant_type", "authorization_code");
+        FormContentProvider formContentProvider = new FormContentProvider(fields);
+        Request request = httpClient.POST(configuration.getTokenEndpoint())
+                .content(formContentProvider)
+                .timeout(10, TimeUnit.SECONDS);
+        ContentResponse response = request.send();
+        String responseBody = response.getContentAsString();
         if (LOG.isDebugEnabled())
-            LOG.debug("claimAuthCode {}", authCode);
-
-        // Use the authorization code to get the id_token from the OpenID Provider
-        String urlParameters = "code=" + authCode +
-            "&client_id=" + UrlEncoded.encodeString(configuration.getClientId(), StandardCharsets.UTF_8) +
-            "&client_secret=" + UrlEncoded.encodeString(configuration.getClientSecret(), StandardCharsets.UTF_8) +
-            "&redirect_uri=" + UrlEncoded.encodeString(redirectUri, StandardCharsets.UTF_8) +
-            "&grant_type=authorization_code";
-
-        URL url = new URL(configuration.getTokenEndpoint());
-        HttpURLConnection connection = (HttpURLConnection)url.openConnection();
-        try
-        {
-            connection.setDoOutput(true);
-            connection.setRequestMethod("POST");
-            connection.setRequestProperty("Host", configuration.getIssuer());
-            connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+            LOG.debug("Authentication response: {}", responseBody);
 
-            try (DataOutputStream wr = new DataOutputStream(connection.getOutputStream()))
-            {
-                wr.write(urlParameters.getBytes(StandardCharsets.UTF_8));
-            }
-
-            try (InputStream content = (InputStream)connection.getContent())
-            {
-                return (Map)JSON.parse(IO.toString(content));
-            }
-        }
-        finally
-        {
-            connection.disconnect();
-        }
+        return (Map)JSON.parse(responseBody);
```

Comment:
This code is dangerous. Please change back to 
```
Object parsedResult = JSON.parse(responseBody);

if (parsedResult instanceof Map)
            {
                Map<?, ?> rawResult = (Map)parsedResult;

                result = rawResult.entrySet().stream().collect(Collectors.toMap(it -> it.getKey().toString(), Map.Entry::getValue));

                LOG.debug("Got result from token server: {}", result);
            }
            else
            {
                LOG.warn("OpenID provider did not return a proper JSON object response. Result was '{}'", responseBody);

                throw new IllegalStateException("Could not parse OpenID provider's malformed response");
            }
```

Reply:
Any error here will be caught in `OpenIdLoginService.login()`, given to `LOG.warn()`, then we return a null `UserIdentity`.

In the above there is a warning then it throws which will eventually be warned again.

Reply:
I don't think this code here is necessary:
```
Map<?, ?> rawResult = (Map)parsedResult;
result = rawResult.entrySet().stream().collect(Collectors.toMap(it -> it.getKey().toString(), Map.Entry::getValue));
```
If it is a map I think we should just be able to cast it to `Map<String, Object>` as it was output by our own JSON parser. I have seen other examples in the code doing similar.
https://github.com/eclipse/jetty.project/blob/20e7aa01f2057299e75302e7370d8b308c20161c/examples/async-rest/async-rest-jar/src/main/java/org/eclipse/jetty/example/asyncrest/AsyncRestServlet.java#L185

Reply:
The double warning isn't good of course. So, that's good and fine about the logging.

Due to type erasure the case isn't safe. You have to take that run-time hit to be type safe. IMO, it's worth it no matter what other parts of the code base are doing. The more the compiler does to help us find bugs, the better.

Reply:
In the places we expect the JSON to be map I am now testing before the cast.